### PR TITLE
Implementation of -u option

### PR DESCRIPTION
**Pull request by ferion11**
Implementation of -u option. Ref: https://github.com/AppImage/zsync2/issues/10
It's work with github links.

* But I found one bug using the kernel.org links:
"zsync_legacy: got non-HTTP response 'HTTP/2 206"

### CODE REVIEW

**Comment by probonopd** [comment]
Is HTTP/2 even supported @TheAssassin? According to https://curl.haxx.se/docs/http2.html it seems like one needs to set `CURLOPT_HTTP_VERSION` to `CURL_HTTP_VERSION_2_0` to support it.

**Comment by ferion11** [comment]
Just to remember, it came from here:
https://github.com/AppImage/zsync2/blob/f41aed234cd790b5d78e420e3fbeb818f8667f58/src/legacy_http.c#L524-L527

File: src/zsclient.cpp [hunk 1]
```diff
@@ -597,7 +598,7 @@ namespace zsync2 {
             // URL might be relative -- we need an absolute URL to do a fetch
             std::string absoluteUrl;
 
-            if (!makeUrlAbsolute(referer, url, absoluteUrl)) {
+            if (!makeUrlAbsolute(referer, url, absoluteUrl) && url_alt.empty()) {
```
Comment (TheAssassin) [inline src/zsclient.cpp]: Not a fan of this if/else construct. Basically, the empty check is repeated twice. It's possible to check only once by wrapping the existing check in another `if`. Please refactor.

File: include/zsclient.h [hunk 2]
```diff
@@ -25,6 +25,10 @@ namespace zsync2 {
         // returns true if a message is available and sets passed string, otherwise returns false
         bool nextStatusMessage(std::string& message);
 
+        // sets URL to get the target file
+        // the server
+        void setUrl(const std::string& url_par);
```
Comment (TheAssassin) [inline include/zsclient.h]: I'd suggest `newUrl`. We're also using camelCase.

File: src/zsclient.cpp [hunk 3]
```diff
@@ -37,6 +37,7 @@ namespace zsync2 {
         // there might be more than one seed file
         // using a set to avoid duplicate entries
         std::set<std::string> seedFiles;
+        std::string url_alt = "";
```
Comment (TheAssassin) [inline src/zsclient.cpp]: We're using camelCase, I'd suggest `userSpecifiedUrl`.

File: include/zsclient.h [hunk 4]
```diff
@@ -25,6 +25,9 @@ namespace zsync2 {
         // returns true if a message is available and sets passed string, otherwise returns false
         bool nextStatusMessage(std::string& message);
 
+        // sets new URL to get the target file from a mirror server
+        void newUrl(const std::string& url);
```
Comment (TheAssassin) [inline include/zsclient.h]: This is a setter, so it should be called `setNewUrl`. See e.g., `setCwd`.

File: src/zsclient.cpp [hunk 5]
```diff
@@ -38,6 +38,9 @@ namespace zsync2 {
         // using a set to avoid duplicate entries
         std::set<std::string> seedFiles;
         
+        std::string userSpecifiedUrl = "";
+        bool userSpecifiedUrlisEmpty = true;
```
Comment (TheAssassin) [inline src/zsclient.cpp]: Why does this have to be an instance variable? It can be queried from the variable above.

File: src/zsclient.cpp [hunk 6]
```diff
@@ -597,13 +600,18 @@ namespace zsync2 {
...
-            if (!makeUrlAbsolute(referer, url, absoluteUrl)) {
-                issueStatusMessage("URL '" + url + "' from .zsync file is relative, which cannot be resolved without "
-                                   "knowing the URL to the .zsync file (you're most likely trying to use a .zsync "
-                                   "file you downloaded from the internet). Without knowing the original URL, it is "
-                                   "impossible to resolve the URL from the .zsync file. Please specify a URL with the "
-                                   "-u flag, or edit and fix the lines in the .zsync file directly.");
-                return -1;
+            if (!userSpecifiedUrlisEmpty) {
+                absoluteUrl = userSpecifiedUrl;
+            }
```
Comment (TheAssassin) [inline src/zsclient.cpp]: Code style. `else` should be on the same line as `}`.

File: src/zsclient.cpp [hunk 7]
```diff
@@ -597,13 +600,18 @@ namespace zsync2 {
...
+            else {
+                        if (!makeUrlAbsolute(referer, url, absoluteUrl)) {
```
Comment (TheAssassin) [inline src/zsclient.cpp]: Code style. We don't use 8 spaces for indentation.

File: src/zsclient.cpp [hunk 8]
```diff
@@ -1047,6 +1055,11 @@ namespace zsync2 {
         }
     };
 
+    void ZSyncClient::newUrl(const std::string& url) {
+        d->userSpecifiedUrl = url;
+        d->userSpecifiedUrlisEmpty = url.empty();
```
Comment (TheAssassin) [inline src/zsclient.cpp]: That variable isn't needed at all.

File: src/zsclient.cpp [hunk 9]
```diff
@@ -750,7 +758,7 @@ namespace zsync2 {
             // copy value for later use
             int okUrls = n;
 
-            if (!url) {
+            if (!url && userSpecifiedUrlisEmpty) {
```
Comment (TheAssassin) [inline src/zsclient.cpp]: The check you added seems superfluent. You're setting `absoluteUrl`, so it should be sufficient to check whether `absoluteUrl` is set to some meaningful value, right?

**Pull request merged**